validator/server_arb/machine_cache.go

```diff
@@ -62,56 +66,64 @@ func NewMachineCache(ctx context.Context, initialMachineGetter func(context.Cont
 		cache.machineStepInterval = config.InitialSteps
 		err = cache.populateInitialCache(ctx, ^uint64(0))
 		if err != nil {
-			cache.SignalReady(err)
+			cache.ProduceError(err)
 			return
 		}
 		cache.finalMachine = cache.machines[len(cache.machines)-1]
 		cache.finalMachine.Freeze()
-		cache.SignalReady(nil)
+		cache.Produce(struct{}{})
 	}()
 	return cache
 }
 
 func (c *MachineCache) SpawnCacheWithLimits(ctx context.Context, start uint64, end uint64) *MachineCache {
-	newInterval := (start - end) / uint64(c.config.TargetMachineCount)
+	newInterval := (start - end) / uint64(c.config.CachedChallengeMachines)
 	if start == c.firstMachineStep && newInterval == c.machineStepInterval {
 		return c
 	}
 	newCache := &MachineCache{
-		ReadyMarker: readymarker.NewReadyMarker(),
-		config:      c.config,
+		Promise: containers.NewPromise[struct{}](),
+		config:  c.config,
 	}
 	go func() {
-		err := c.WaitReady(ctx)
+		_, err := c.Await(ctx)
 		if err != nil {
-			newCache.SignalReady(err)
+			newCache.ProduceError(err)
 			return
 		}
 		newCache.zeroStepMachine = c.zeroStepMachine
 		newCache.finalMachine = c.finalMachine
 		closest, err := c.getClosestMachine(start)
```

Comment:
Theoretically this could use `lastMachine` but doesn't, but it's probably not the end of the world.

Reply:
you're right. In current way it will be hard, but I want a future version with some more locking and there we should also try to use last_machine